packages/eslint-plugin/src/rules/no-unnecessary-type-parameters.ts

```diff
@@ -303,7 +303,7 @@ function collectTypeParameterUsageCounts(
         if (properties.length === 0) {
           // TS treats mapped types like `{[k in "a"]: T}` like `{a: T}`.
           // They have properties, so we need to avoid double-counting.
-          visitType(type.templateType, false);
+          visitType(type.templateType ?? type.constraintType, false);
```

Comment:
From https://github.com/typescript-eslint/typescript-eslint/pull/9530:

> The rule previously descended into the typeParameter. It doesn't need to descend into the constraintType because we get that through another code path. It does need to descend into the templateType, though.

I wish I'd included more details about what "another code path" was and what my example was. Since you touched this more recently, do you have an idea? Double-counting would be problematic for this rule since it would lead to false negatives. On the other hand, if we only require one appearance of `T` in the return slot, then we can double-count all we want.

I'd also like to double-check whether this comment has any implications for this change:

> One wrinkle is that if the mapped type is something like {[k in 'a']: T}, then TypeScript will treat it like {a: T} and we'll process it as an object type. In this case, we need to avoid also processing it as a mapped type, even though isMappedType returns true.

Reply:
I ... don't 😅. I couldn't figure out how to construct a test case that shows a flaw in this changed logic.

Same with the mapped types like `{ [k in 'a'] ... }`.